vamb/hier.py

```diff
@@ -0,0 +1,480 @@
+    # TODO: Could consider removing for non-rooted sub-trees?
+    assert np.all(np.count_nonzero(subset_paths >= 0, axis=1) - 1 == deepest)
+    return subset_paths[np.arange(tree.num_nodes()), deepest]
+
+
+def _last_nonzero(x, axis):
+    x = np.asarray(x, bool)
+    assert np.all(np.any(x, axis=axis)), 'at least one must be nonzero'
+    # Find last element that is true.
+    # (First element that is true in reversed array.)
+    n = x.shape[axis]
+    return (n - 1) - np.argmax(np.flip(x, axis), axis=axis)
+
+
+def uniform_leaf(tree: Hierarchy) -> np.ndarray:
+    """Returns a uniform distribution over leaf nodes."""
+    is_ancestor = tree.ancestor_mask(strict=False)
+    is_leaf = tree.leaf_mask()
+    num_leaf_descendants = is_ancestor[:, is_leaf].sum(axis=1)
+    return num_leaf_descendants / is_leaf.sum()
+
+
+def uniform_cond(tree: Hierarchy) -> np.ndarray:
+    """Returns a uniform distribution over child nodes at each conditional."""
+    # TODO: Ensure exact.
+    node_to_num_children = {k: len(v) for k, v in tree.children().items()}
+    num_children = np.asarray([node_to_num_children.get(x, 0) for x in range(tree.num_nodes())])
+    parent_index = tree.parents()
+    # Root node has likelihood 1 and no parent.
+    log_cond_p = np.concatenate([[0.], -np.log(num_children[parent_index[1:]])])
+    is_ancestor = tree.ancestor_mask(strict=False)
+    log_p = np.dot(is_ancestor.T, log_cond_p)
+    return np.exp(log_p)
+
+
+def lca_depth(tree: Hierarchy, inds_a: np.ndarray, inds_b: np.ndarray) -> np.ndarray:
+    """Returns the depth of the LCA node.
+
+    Supports multi-dimensional index arrays.
+    """
+    paths = tree.paths_padded(exclude_root=True)
+    paths_a = paths[inds_a]
+    paths_b = paths[inds_b]
+    return np.count_nonzero(
+        ((paths_a == paths_b) & (paths_a >= 0) & (paths_b >= 0)),
+        axis=-1)
+
+
+def find_lca(tree: Hierarchy, inds_a: np.ndarray, inds_b: np.ndarray) -> np.ndarray:
+    """Returns the index of the LCA node.
+
+    Supports multi-dimensional index arrays.
+    For example, to obtain an exhaustive table:
+        n = tree.num_nodes()
+        find_lca(tree, np.arange(n)[:, np.newaxis], np.arange(n)[np.newaxis, :])
+    """
+    paths = tree.paths_padded(exclude_root=False)
+    paths_a = paths[inds_a]
+    paths_b = paths[inds_b]
+    num_common = np.count_nonzero(
+        ((paths_a == paths_b) & (paths_a >= 0) & (paths_b >= 0)),
+        axis=-1)
+    return paths[inds_a, num_common - 1]
+
+
+class FindLCA:
+
+    def __init__(self, tree: Hierarchy):
+        self.paths = tree.paths_padded(exclude_root=False)
+
+    def __call__(self, inds_a: np.ndarray, inds_b: np.ndarray) -> np.ndarray:
+        paths = self.paths
+        paths_a = paths[inds_a]
+        paths_b = paths[inds_b]
+        num_common = np.count_nonzero(
+            ((paths_a == paths_b) & (paths_a >= 0) & (paths_b >= 0)),
+            axis=-1)
+        return paths[inds_a, num_common - 1]
```

Comment:
Not that I'm against it, but why make a class just for one method? Wouldn't a function be easier?

Reply:
those are some leftovers from the code from another repo, might look into this after the pipeline is sorted